src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.CoreCLR.cs

```diff
@@ -49,81 +50,65 @@ public void Pop()
         // that we should continue on.
-        CORINFO_CONTINUATION_CONTINUE_ON_CAPTURED_TASK_SCHEDULER = 32,
+        CORINFO_CONTINUATION_CONTINUE_ON_CAPTURED_TASK_SCHEDULER = 64,
     }
 
-    internal sealed unsafe class Continuation
+#pragma warning disable CA1852 // "Type can be sealed" -- no it cannot because the runtime constructs subtypes dynamically
+    internal unsafe class Continuation
     {
         public Continuation? Next;
-        public delegate*<Continuation, Continuation?> Resume;
-        public uint State;
+        public delegate*<Continuation, ref byte, Continuation?> Resume;
         public CorInfoContinuationFlags Flags;
+        public int State;
 
-        // Data and GCData contain the state of the continuation.
-        // Note: The JIT is ultimately responsible for laying out these arrays.
-        // However, other parts of the system depend on the layout to
-        // know where to locate or place various pieces of data:
-        //
-        // 1. Resumption stubs need to know where to place the return value
-        // inside the next continuation. If the return value has GC references
-        // then it is boxed and placed at GCData[0]; otherwise, it is placed
-        // inside Data at offset 0 if
-        // CORINFO_CONTINUATION_OSR_IL_OFFSET_IN_DATA is NOT set and otherwise
-        // at offset 4.
-        //
-        // 2. Likewise, Finalize[Value]TaskReturningThunk needs to know from
-        // where to extract the return value.
-        //
-        // 3. The dispatcher needs to know where to place the exception inside
-        // the next continuation with a handler. Continuations with handlers
-        // have CORINFO_CONTINUATION_NEEDS_EXCEPTION set. The exception is
-        // placed at GCData[0] if CORINFO_CONTINUATION_RESULT_IN_GCDATA is NOT
-        // set, and otherwise at GCData[1].
-        //
-        public byte[]? Data;
-        public object?[]? GCData;
-
-        public object GetContinuationContext()
+#if TARGET_64BIT
+        private const int POINTER_SIZE = 8;
+#else
+        private const int POINTER_SIZE = 4;
+#endif
+
+        private const int CONTINUATION_OFFSET_TO_DATA = POINTER_SIZE /* Next */ + POINTER_SIZE /* Resume */ + 8 /* Flags + State */;
+
+        public unsafe object GetContinuationContext()
         {
-            int index = 0;
-            if ((Flags & CorInfoContinuationFlags.CORINFO_CONTINUATION_RESULT_IN_GCDATA) != 0)
-                index++;
-            if ((Flags & CorInfoContinuationFlags.CORINFO_CONTINUATION_NEEDS_EXCEPTION) != 0)
-                index++;
-            Debug.Assert(GCData != null && GCData.Length > index);
-            object? continuationContext = GCData[index];
-            Debug.Assert(continuationContext != null);
-            return continuationContext;
+            Debug.Assert((Flags & CorInfoContinuationFlags.CORINFO_CONTINUATION_HAS_CONTINUATION_CONTEXT) != 0);
```

Comment:
(preexisting problem)

`CORINFO_...` types and constants are meant to be exchange types on JIT/EE interface and they are not meant to leak to runtime. It is odd to see them in CoreLib code.

(When these types leak to runtime, it often causes confusion. #116675 fixed one of such leaks that caused confusion for interpreter v-team.)

Reply:
Is it strictly the name of the enum/enum members on the C# side you are thinking about, or do you also think the values of the enum members should be communicated to the JIT by the VM in a different way?